### DISCUSSION TITLE
Starfield inspired re-style

### DISCUSSION THREAD

**Discussion by RSchwoerer** [aquasarus/Starfield-Interactive-Smart-Slate#11] [Show and tell]
Hey @aquasarus ! This is a really cool idea and app!
I have been thinking about something similar, I cannot believe this doesn't exist in-game.

I am a C# and WPF developer IRL, and I thought I would try to give the app a style that fits with the game.
Below is just a starting point re-styling the celestial body panel. I have some other ideas for other parts of the app as well.
Let me know what you think and if you're open to some help, if not no worries.

⚠️warning: work in progress⚠️

![image](https://github.com/aquasarus/Starfield-Interactive-Smart-Slate/assets/5709991/b98af92c-46f7-4e40-bcad-124c88642b98)

**Comment by DeHades666** [comment]
Damn.. All you clever people... Great idea =)

**Comment by aquasarus** [comment]
@RSchwoerer Well I'll be damned. A real WPF dev! And you made like 2000 line changes just to show me a demo! It looks _amazing_. Here's some background and my thoughts:
- First of all, I always imagined myself tapping away at a data slate in-game while using this app. Because of this, my had this stretch goal to build a UI on top of a data slate background. Even though the available screen space would be smaller due to the data slate aspect ratio (in both landscape and portrait), I feel like it would be so cool and immersive to use that as a companion app.
- To build a UI like that would require significant time investment from me, so it would only be worth it if there are enough people who use it. I haven't really looked into the details of actually doing it. This is where your demo comes in. Having a dark theme that matches the in-game UI style will already help a lot with immersion. I was also mimicking the in-game planetary overview when I built this, so it's such a pleasant surprise to see you replicate it with this level of accuracy!
- That said, my main reservation against jumping headfirst into this is the issue of iterability. If I build a UI into this app that looks THIS good, I'd want it to stay this good. That means anything new that I add will have to go through the same process. The core idea of this app had always been about function over form. I wouldn't want to feel like I could quickly add a very useful feature to the app, only to be slowed down by building a full-fledged UI that matches the theme.
- So I'm thinking...could there be a lower effort "dark theme" that can be toggled on? I was disappointed to find that WPF didn't have a built-in dark theme, but I imagine recoloring the UI controls wouldn't add _too much_ overhead. This would allow the app to have a color-palette that more closely resembles the game, while still preserving the vanilla version of this app. Building a new control will also just involve deciding colors for the dark version. 

Let me know what you think! This demo has been incredible to witness. This is also my first time building a WPF app, so I'm sure having a veteran to work with will go a long way :).

**Reply by RSchwoerer** [reply]
This has been a fun project so far, I love the game and the aesthetic and recreating it is a fun design challenge.
 
> And you made like 2000 line changes
Ha, some things in WPF are _very_ verbose, and some things are unbelievably trivial. It is an interesting framework. 

The data slate idea theming would be a cool look, but I agree I'm not sure how that would affect the usability. Also looking through the current use in-game its not clear to me how much interaction is possible on them, are they more like a kindle? I think we can incorporate it somehow, but I'm not sure what that looks like at the moment.

Regarding the rest of the UI styles, once some basic control styles and defaults are established it is trivial to use those throughout the app. Most styles can even be set as default, so they will used automatically. Things like colors, fonts, even margin values can all be stored as constants and used throughout all of the control styles. True there is some initial effort to determine the styles for each of the controls being used, but once that is done it is much easier to just use the control as normal and get the style basically for free.

WPF does have theme support, although it is a bit weird to set up and is not really documented. As an example there are a number of WPF control libraries that support themes, often dynamic theme switching. That being said, I'm not sure how important theme switching support is, my thought was simply to resemble the existing game styles as much as is practical.

If you think theme swapping would be an important addition we can look into adding it. Themes require every color and brush to be a resource constant so that they can be swapped out dynamically, so every control needs to be restyled to use those resources anyway.
It is less work (and would look better IMHO) to simply restyle the controls as we need them than it would be to try to make a generic dark theme for the default control styles.


So, my recommendation...
- Continue to design and apply controls styles for the existing controls in use with a Starfield design flavor.
- Figure out how and where to add more visual flair where practical, i.e. data slate or Starware elements from the game while maintaining your design goal of "this app had always been about function over form".

I have some other ideas for the app also, not sure where those should go, maybe a new discussion thread for roadmap items?

**Reply by aquasarus** [reply]
Hmm I see your point. And I do agree that if we can build a good style without compromising on iterability, there is no reason to support multiple themes. I'll trust your judgement on the feasibility here. If you're serious about helping out for this app, and it seems like you are, let's talk about some specifics:
1. I've been taking advantage of the app being a one-man show to bring it together quickly. That, combined with my inexperience in WPF, means a lot of the code is either going to be inefficient for collaboration, or just plain wrong from a best-practices standpoint. If we're going to work together on this, we need to come up with a way to work on a full UI revamp while still allowing me to upgrade its functionality. To me, this means properly splitting out UI from business logic. Unless you think we can afford to focus entirely on the UI before touching anything else, I'd like your advice on how I should restructure this codebase in terms of its view vs model setup.
1. I've also been planning on properly consolidating all string resources instead of the current, hardcoded mess. I've been using ChatGPT 3.5 to help me build WPF stuff without any prior knowledge, which works but is far from ideal. So perhaps you can shed some light on the most modern way to localize a WPF app? I'd like to support other languages in the near future, and @birilo has offered to translate to French for us!
1. Yes, by all means start as many discussions as you'd like! That, or we can switch to something like Discord to chat more actively.
1. Lastly, as it seems like we are on the verge of starting a collaboration, I hope you'll understand that I may not be able to work on this as quickly as you might like. I'm a full-time parent and I only have 3-4 hours to myself on a daily basis. Nonetheless, I am excited to see where we can take this :).

**Reply by RSchwoerer** [reply]
I am also a parent, and have a full-time job, so this squarely falls into hobby territory for me.
And it cannot cut into my Starfield playing time 🤣.
I have no expectations about velocity, I'm just offering to provide some help and hope to make a cool thing. Feel free to tell me if something doesn't align with your vision of your app.


> combined with my inexperience in WPF, means a lot of the code is either going to be inefficient for collaboration, or just plain wrong from a best-practices standpoint

While there are things I would have done differently, the current code certainly isn't _wrong_ in any regard. One huge reason I've been working on this in my free time is the informal-ness of it. The current code gets the job done.

I acknowledge your comment about collaboration. There are a few quick changes that can be made to break up the code (I'm looking at you MainWindow.xaml.cs). I would start moving to more of an MVVM architecture, that will split things up further, and should simplify some of the existing code-behind logic. This can be done over time also, we don't necessarily have to change everything at once.
Depending on what the additional new functionality you have in mind is, I don't see too significant an issue with more than one person working on it.

> been using ChatGPT 3.5 to help me build WPF stuff without any prior knowledge

This is pretty amazing. I haven't used it for anything incredibly useful.

Regarding localization, we use our own custom framework at work.... I looked briefly into using something more simple, it doesn't seem too complex to add, a lot if not all of the functionality is there already in the framework. 


If you want to set up a Discord server maybe that would be ideal for quicker discussion, could be another venue for bug reporting and feedback also (so people don't need a GH login), I don't mind either way.

**Reply by aquasarus** [reply]
> And it cannot cut into my Starfield playing time 🤣.

Haha I got hooked _hard_ with this game. 200 hours later, now I just picture myself working in the MAST building to develop the next version of Starware data slates :P.

Good to know that the codebase isn't too out of shape. Seems like everything is moving towards MVVM these days, so I'll start looking into fixing things up as I go. I don't have any big plans for changes at the moment. Here's a quick list of relatively uncomplicated changes I want to make in the near future:
- Searching for multiple resources simultaneously. (Should be a super straightforward list multi-select.)
- Extending current pictures/notes functionality to planets and moons. (Leveraging existing architecture, although this will likely be put under the current celestial body overview. So we'll have to keep it in mind for the UI refresh.)
- The beginning of French localization. (The UI translation part is actually secondary. I mostly want to allow Fauna/Flora to have French auto-complete. And @birilo has already linked me a French wiki with this information.)
- Integrating the "farmable" attribute of lifeforms into the app. (A new flag in the DB and a checkbox in the lifeform edit dialog.)
- MAYBE Traits. This is a longer-term goal because I haven't even gathered a comprehensive list of Traits yet. Plus, we'll probably want to get the icons for each to display in our app, so the non-code workload is already substantial. On the other hand, I feel like it isn't something I can omit if I want the app to be a compendium/codex.

Regarding ChatGPT, I should clarify that I'm still a software engineer by trade. ChatGPT just helps me get on the right track with new tools. It also helps reduce some of the busywork involved in dev (e.g. ask it to write regex for you!).

I made a Discord server but we'll keep it dev-only for now. I think our users have enough channels to submit issues for the time being (most of them probably don't leave the [Nexus project page](https://www.nexusmods.com/starfield/mods/7074?tab=description&BH=0)). Send me a Discord DM on @aquasarus and I'll get you the invite.
